### PR TITLE
feat: hover providers for refs and sources

### LINKED ISSUE #525: Include "type signature" from the schema.yml docs in the ref preview
### Describe the feature

The current command+hover preview looks like:
Screenshot 2023-08-14 at 2 14 15 PM

It includes the config block from the model. It would be nice to see some more metadata in there like the table description or even the columns if it could be presented in a readable way.

### Describe alternatives you've considered

You can kind of get to this with an extra click if you go-to the definition of the ref and read the docs but then you are opening a new file.

### Who will benefit?

This is a pretty generalizable feature. I suspect that most folks don't use the preview as much as they just click through to the definition.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by jsnb-devoted**
## Overview
Starts to resolve #525
This is a real copy paste paint by numbers kind of job. I'm modeling all this code on how the definition provider is done. There is likely some code that is worth abstracting here and the hover itself isn't the prettiest but I thought this was a good start. Happy to keep iterating on this if folks want a more full featured version of this.

Additional features this could use:
1. Parse tests from the manifest for the refs so you can show tests
2. Parse descriptions/columns/types from sources
3. Do a macro hover provider too.
4. Better styling on the markdown

I tested this on a jaffle shop repo. Here are some screenshots:
Screenshot 2023-09-16 at 2 46 54 PM
Screenshot 2023-09-16 at 2 46 58 PM
Screenshot 2023-09-16 at 2 47 02 PM

## Checklist

- [x] I have run this code and it appears to resolve the stated issue
- [ ] `README.md` updated and added information about my change

### CODE REVIEW

File: src/hover_provider/modelHoverProvider.ts [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+      content.appendMarkdown("---");
+      content.appendText("\n");
+      content.appendText("\n");
+      for (const colKey in node.columns) {
+        const column = node.columns[colKey];
+        content.appendMarkdown(
+          `<br/><span style="color:#347890;">(column)&nbsp;</span><span>${column.name} &nbsp;</span>`,
+        );
+        if (column.data_type !== null) {
+          content.appendMarkdown(`<span>${column.data_type}&nbsp;</span>`);
```
Comment (mdesmet) [inline src/hover_provider/modelHoverProvider.ts:146]: Could you explain in which cases the `data_type` is available?

File: src/hover_provider/modelHoverProvider.ts [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+      content.appendText("\n");
+      content.appendMarkdown("---");
+      content.appendText("\n");
+      content.appendText("\n");
+      for (const colKey in node.columns) {
+        const column = node.columns[colKey];
+        content.appendMarkdown(
+          `<br/><span style="color:#347890;">(column)&nbsp;</span><span>${column.name} &nbsp;</span>`,
+        );
+        if (column.data_type !== null) {
```
Comment (mdesmet) [inline src/hover_provider/modelHoverProvider.ts:145]: Could you make this view more compact? What are the limits? How many lines can be displayed. Is there any way to expand it?

File: src/hover_provider/modelHoverProvider.ts [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+  ): ProviderResult<Hover> {
+    return new Promise((resolve) => {
+      const hover = document.getText(document.getWordRangeAtPosition(position));
+      const word = document.getText(
+        document.getWordRangeAtPosition(position, ModelHoverProvider.IS_REF),
+      );
+      const project = this.dbtProjectContainer.findDBTProject(document.uri);
+      if (!project) {
+        console.error(
+          "Could not load definition provider, project not found in container for " +
```
Comment (mdesmet) [inline src/hover_provider/modelHoverProvider.ts:59]: Please adapt the error description.

File: src/hover_provider/modelHoverProvider.ts [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+          const mdString = this.getDefinitionFor(
+            project.getProjectName(),
+            dbtModel[0],
+            document.uri,
+          );
+          if (mdString !== undefined) {
+            const hover = new Hover(mdString, new Range(position, position));
+            resolve(hover);
+          }
+          this.telemetry.sendTelemetryEvent("provideModelDefinition", {
```
Comment (mdesmet) [inline src/hover_provider/modelHoverProvider.ts:76]: Please adapt the telemtry event name to `provideModelHover`

File: src/hover_provider/modelHoverProvider.ts [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+          const mdString = this.getDefinitionFor(
+            dbtModel[0],
+            dbtModel[2],
+            document.uri,
+          );
+          if (mdString !== undefined) {
+            const hover = new Hover(mdString, new Range(position, position));
+            resolve(hover);
+          }
+          this.telemetry.sendTelemetryEvent("provideModelDefinition", {
```
Comment (mdesmet) [inline src/hover_provider/modelHoverProvider.ts:91]: See above.

File: src/hover_provider/modelHoverProvider.ts (hunk 1 above)
Comment (jsnb-devoted) [reply src/hover_provider/modelHoverProvider.ts:146]: It's part of the standard manifest schema:
https://docs.getdbt.com/reference/resource-properties/columns

File: src/hover_provider/modelHoverProvider.ts (hunk 2 above)
Comment (jsnb-devoted) [reply src/hover_provider/modelHoverProvider.ts:145]: I'll have to look into this. It does scrolling by default. The first screen shot in the description has scrolling enabled.

Comment (jsnb-devoted) [reply src/hover_provider/modelHoverProvider.ts:145]: Uploading Screen Recording 2023-09-17 at 11.20.41 AM.mov…

File: src/hover_provider/modelHoverProvider.ts (hunk 1 above)
Comment (jsnb-devoted) [reply src/hover_provider/modelHoverProvider.ts:146]: Screenshot 2023-09-17 at 11 33 26 AM
Screenshot 2023-09-17 at 11 34 01 AM

File: src/hover_provider/modelHoverProvider.ts (hunk 2 above)
Comment (jsnb-devoted) [reply src/hover_provider/modelHoverProvider.ts:145]: https://github.com/AltimateAI/vscode-dbt-power-user/assets/79863718/e9e7abd4-8277-4abc-8e68-721bb4c543d6

Comment (jsnb-devoted) [reply src/hover_provider/modelHoverProvider.ts:145]: It actually fits the screen width:
https://github.com/AltimateAI/vscode-dbt-power-user/assets/79863718/15702629-c9a9-46ab-a656-a0376a1ccfcf

File: src/hover_provider/modelHoverProvider.ts (hunk 3 above)
Comment (jsnb-devoted) [reply src/hover_provider/modelHoverProvider.ts]: sorry about all the copy/pasting here.

**Comment by mdesmet** [comment]
@jsnb-devoted : Can you please drop the merge commit and rebase on master instead?

File: src/hover_provider/modelHoverProvider.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+          content.appendMarkdown(
+            `<span>-&nbsp;${column.data_type.toUpperCase()}</span>`,
+          );
+        }
+        if (column.description !== null) {
+          content.appendMarkdown(
+            `<br/><span><em>${column.description}</em></span>`,
+          );
+        }
+      }
```
Comment (mdesmet) [inline src/hover_provider/modelHoverProvider.ts:155]: When no description is provided, the spacing is a bit extraneous. Could you reduce it so it has the same spacing compared to the case when description is there.

Comment (jsnb-devoted) [reply src/hover_provider/modelHoverProvider.ts]: I thought I had handled that in a recent commit but I was checking for `null` instead of empty strings in the descriptions. Looks a lot better without the spacing now.
Screenshot 2023-09-19 at 7 26 43 PM
Screenshot 2023-09-19 at 7 27 20 PM

**Comment by anandgupta42** [comment]
@jsnb-devoted Since @mdesmet is reviewing the PR, I spent some time to smoke-test the PR. 

I used https://github.com/duneanalytics/spellbook as the repository for testing, and everything seems to be working well with the reference. However, hovering over sources did not work. Additionally, the text sometimes overlapped with the scroll bar, which made the reading experience less than optimal. I have attached a couple of screenshots as evidence of these issues. Perhaps we can improve the user experience by adding some extra space to the layout.


Screenshot 2023-09-20 at 9 41 21 AM
Screenshot 2023-09-20 at 9 36 24 AM

**Comment by jsnb-devoted** [comment]
> @jsnb-devoted Since @mdesmet is reviewing the PR, I spent some time to smoke-test the PR.

Thanks @anandgupta42. I only have so much time to test the edge cases. I appreciate you going thru these.
 
> I used https://github.com/duneanalytics/spellbook as the repository for testing, and everything seems to be working well with the reference. However, hovering over sources did not work. 

I couldn't recreate this with the spellbook project with the standard use of the `source` macro.
Screenshot 2023-09-20 at 10 14 20 AM

I did notice it doesn't work in a macro where the first argument is a variable instead of a string like:
```
{% macro 
    oneinch_ar_calls_transfers_macro(
        blockchain,
        project_start_date_str,
        wrapper_token_address
    ) 
%}
...
SELECT *
from {{ source(blockchain, 'transactions') }}
```
([github](https://github.com/duneanalytics/spellbook/blob/main/macros/models/_project/oneinch/oneinch_ar_calls_transfers_macro.sql#L56))
off the top of my head I'm not sure it makes sense to support that but you can also see that the goto definition provider doesn't work for that case either. I would be fine not including that in the scope of this feature but it's up to you all.

What source did you observe the hover not showing up?


> Perhaps we can improve the user experience by adding some extra space to the layout.

This seems pretty doable. I'm not terribly UI inclined and I didn't even notice the overlap because it is pretty small. I suspect I can add a little padding to sort that out.

**Comment by jsnb-devoted** [comment]
>Additionally, the text sometimes overlapped with the scroll bar, which made the reading experience less than optimal. I have attached a couple of screenshots as evidence of these issues. Perhaps we can improve the user experience by adding some extra space to the layout

@anandgupta42 -- At first glance I can't figure out how to add the extra space here. The markdown string supports limited html and style ([stackoverflow](https://stackoverflow.com/a/67954180/7584305)). I was able to change the color of the text inside the description span but I can't seem to change the padding.

**Comment by anandgupta42** [comment]
> This seems pretty doable. I'm not terribly UI inclined and I didn't even notice the overlap because it is pretty small. I suspect I can add a little padding to sort that out.

Yes. That should do the trick.

> I couldn't recreate this with the spellbook project with the standard use of the `source` macro.
Sorry about this. I meant macros. Not sure why I ended up writing sources.

**Comment by anandgupta42** [comment]
> @anandgupta42 -- At first glance I can't figure out how to add the extra space here. The markdown string supports limited html and style ([stackoverflow](https://stackoverflow.com/a/67954180/7584305)). I was able to change the color of the text inside the description span but I can't seem to change the padding.

This is not a major issue for me. It would be nice to have, but it's not essential.

**Comment by anandgupta42** [comment]
> Support for `refs` is high enough value IMO

💯

Once @mdesmet approves the PR, we should be good to go.

File: src/hover_provider/modelHoverProvider.ts [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+      const project = this.dbtProjectContainer.findDBTProject(document.uri);
+      if (!project) {
+        console.error(
+          "Could not load hover provider, project not found in container for " +
+            document.uri.fsPath,
+        );
+        return;
+      }
+      if (word !== undefined && hover !== "ref") {
+        const dbtModel = word.match(ModelHoverProvider.GET_DBT_MODEL);
```
Comment (mdesmet) [inline src/hover_provider/modelHoverProvider.ts:65]: Is it possible to extract this logic to a shared utils class?

This would make it easier to evolve this code and fix issues in both the definition provider and the hover provider.

**Review by mdesmet — COMMENTED** [review]
Looks good except the code duplication. Could you have a look at that and check if there is any opportunity for extraction?

File: src/hover_provider/modelHoverProvider.ts (hunk 7 above)
Comment (jsnb-devoted) [reply src/hover_provider/modelHoverProvider.ts:65]: Ya - there is definitely code duplication here. I was hoping to limit this to a v1 sort of feature but I can pull some of this out into shared utility.

Comment (mdesmet) [reply src/hover_provider/modelHoverProvider.ts:65]: It was just a question, let's do this as a follow up.

**Review by mdesmet — APPROVED** [review]
Thanks for this great contribution
